feat: support map data type in lance format version 2.2

xloya opened the pull request:
Close https://github.com/lance-format/lance/issues/3620.
Currently, Lance does not support the `Map` data type. Importing Lance from a data source that supports `Map` data type requires special handling, which incurs significant processing costs, even though this type is very common in other data sources.

This PR aligns with the Map data type in Arrow, implementing the Map logical data type. In actual encoder, it uses an `Offsets Array + List<Struct<key, value>>` approach, which is similar to `List` data type .
And in actual decoder, it will decode the struct array and offset infos to restruct to the Arrow `MapArray`.

xloya commented:
@westonpace @jackye1995 @Xuanwo @eddyxu PTAL when you have time, thanks!

Xuanwo commented:
> @Xuanwo Thanks for your review! Currently, most of the logic for adding the Map data type is located in the Field, Encoder, and Decoder modules. These modules generally don't contain format version information. I'm unsure if it's appropriate to limit support for the Map data type to version 2.2+, since version 2.1+ uses the same logic for reading, writing, and encoding/decoding (different from version 2.0 and below). Do you have any suggestions on this?

Take Lance v1.0.0 as an example. If we add the `Map` type, users might find that data written in Lance v1.0.0 is not readable by Lance v0.39.0, and that's what we want to avoid.

xloya commented on rust/lance-core/src/datatypes/field.rs:
```diff
@@ -980,16 +992,42 @@ impl TryFrom<&ArrowField> for Field {
+                // TODO: We only support keys_sorted=true for now,
+                //  because converting a rust arrow map field to the python arrow field will
+                //  lose the keys_sorted property.
+                if *keys_sorted {
```
@Xuanwo I tried to restore the `keys_sorted` parameter in the Map type, which worked fine in Rust, but the parameter was lost once it converted to Python Arrow Schema using `PyArrowType`. I think this might be a bug with the Arrow code itself. Therefore, I want to only support Map type with `keys_sorted=false` in this PR, which is also the default value of Arrow MapType, and continue to investigate this issue. I will consider supporting Map type with `keys_sorted=true` in a future PR. But to be honest, this parameter does not actually sort the Map in Arrow; it is only used as metadata, so I don't think it will make a significant difference to users.

Xuanwo replied:
Got it, thank you for working on this

Xuanwo commented on rust/lance-encoding/src/decoder.rs:
```diff
@@ -501,6 +504,20 @@ impl CoreFieldDecoderStrategy {
+    }
+
+    /// Create a new strategy from decoder config with file version
+    pub fn from_decoder_config_with_version(
```
Can we have an `with_file_version` API instead?

xloya commented:
> Here some nits about this PR. Maybe @westonpace wanna take an other look.

@Xuanwo Addressed all the comments, PTAL again when you have time, thanks!

Xuanwo reviewed (approved):
Thank you for working on this! Can you help update our docs as a follow-up?

westonpace reviewed (commented):
This is really good!  Sorry I took a while look at it.  Thanks for the addition :rocket:
